controller/share_manager_controller.go

```diff
@@ -1050,6 +1045,81 @@ func (c *ShareManagerController) getShareManagerTolerationsFromStorageClass(sc *
 	return tolerations
 }
 
+func (c *ShareManagerController) validateStorageNetworkForRWXVolume() (bool, error) {
+	storageNetwork, err := c.ds.GetSettingWithAutoFillingRO(types.SettingNameStorageNetwork)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get setting value %v", types.SettingNameStorageNetwork)
+	}
+
+	storageNetworkForRWXVolumeEnabled, err := c.ds.GetSettingAsBool(types.SettingNameStorageNetworkForRWXVolumeEnabled)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get setting value %v", types.SettingNameStorageNetworkForRWXVolumeEnabled)
+	}
+
+	return types.IsStorageNetworkForRWXVolume(storageNetwork, storageNetworkForRWXVolumeEnabled), nil
+}
+
+func (c *ShareManagerController) checkStorageNetworkApplied() (bool, error) {
+	targetSettings := []types.SettingName{types.SettingNameStorageNetwork, types.SettingNameStorageNetworkForRWXVolumeEnabled}
+	for _, item := range targetSettings {
+		if applied, err := c.ds.GetSettingApplied(item); err != nil || !applied {
+			return applied, err
+		}
+	}
+	return true, nil
+}
+
+func (c *ShareManagerController) cleanupServiceAndEndpoint(shareManager *longhorn.ShareManager) error {
+	service, err := c.ds.GetService(c.namespace, shareManager.Name)
+	if err != nil {
+		// if NotFound, means the service/endpoint is already cleaned up
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to get service for share manager %v", shareManager.Name)
+	}
+
+	applied, err := c.checkStorageNetworkApplied()
+	if err != nil {
+		return errors.Wrap(err, "failed to check if the storage network settings are applied")
+	}
+	if !applied {
+		c.logger.Warn("Storage network settings are not applied, do nothing")
+		return nil
+	}
+
+	storageNetworkForRWXVolumeValidate, err := c.validateStorageNetworkForRWXVolume()
+	if err != nil {
+		return err
+	}
+
+	// no need to cleanup because looks the service file is correct
+	if storageNetworkForRWXVolumeValidate {
+		if service.Spec.ClusterIP == core.ClusterIPNone {
+			return nil
+		}
+	} else {
+		if service.Spec.ClusterIP != core.ClusterIPNone {
+			return nil
+		}
+	}
+
+	// let's cleanup
+	c.logger.Infof("Deleting Service for sharemanager %v", shareManager.Name)
+	err = c.ds.DeleteService(c.namespace, shareManager.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
```suggestion
		return errors.Wrapf(err, "failed to delete endpoint for share manager %v", shareManager.Name)
```

Reply:
Hi @derekbit, I am not sure of this.
We are trying to delete the service in this section, so the error msg should be `failed to delete service` instead of `failed to delete endpoint`. Or did I get some misunderstanding?

Reply:
oh, typo. should be `service`